runtime/src/status_cache.rs

```diff
@@ -122,186 +162,361 @@ impl<T: Serialize + Clone> StatusCache<T> {
+        let mut key_slice = MaybeUninit::<[u8; CACHED_KEY_SIZE]>::uninit();
 
         // Get the cache entry for this blockhash.
-        let (max_slot, key_index, hash_map) =
-            self.cache.entry(*transaction_blockhash).or_insert_with(|| {
+        let key_index = {
+            let (max_slot, key_index, txs) = &*self.cache.get_or_insert_with(blockhash, || {
                 let key_index = thread_rng().gen_range(0..max_key_index + 1);
-                (slot, key_index, HashMap::new())
+                (
+                    AtomicU64::new(slot),
+                    key_index,
+                    DashMap::with_hasher_and_shard_amount(AHashRandomState::default(), KEY_SHARDS),
+                )
             });
 
-        // Update the max slot observed to contain txs using this blockhash.
-        *max_slot = std::cmp::max(slot, *max_slot);
+            // Update the max slot observed to contain txs using this blockhash.
+            max_slot.fetch_max(slot, Ordering::Relaxed);
 
-        // Grab the key slice.
-        let key_index = (*key_index).min(max_key_index);
-        let mut key_slice = [0u8; CACHED_KEY_SIZE];
-        key_slice.clone_from_slice(&key.as_ref()[key_index..key_index + CACHED_KEY_SIZE]);
+            // Grab the key slice.
+            let key_index = (*key_index).min(max_key_index);
+            unsafe {
+                ptr::copy_nonoverlapping(
+                    key.as_ref()[key_index..key_index + CACHED_KEY_SIZE].as_ptr(),
+                    key_slice.as_mut_ptr() as *mut u8,
+                    CACHED_KEY_SIZE,
+                )
+            }
+
+            // Insert the slot and tx result into the cache entry associated with
+            // this blockhash and keyslice.
+            let mut forks = txs.entry(unsafe { key_slice.assume_init() }).or_default();
+            forks.push((slot, res.clone()));
 
-        // Insert the slot and tx result into the cache entry associated with
-        // this blockhash and keyslice.
-        let forks = hash_map.entry(key_slice).or_default();
-        forks.push((slot, res.clone()));
+            key_index
+        };
 
-        self.add_to_slot_delta(transaction_blockhash, slot, key_index, key_slice, res);
+        self.add_to_slot_delta(
+            blockhash,
+            slot,
+            key_index,
+            unsafe { key_slice.assume_init() },
+            res,
+        );
     }
 
-    pub fn purge_roots(&mut self) {
+    fn purge_roots(&self) {
         if self.roots.len() > MAX_CACHE_ENTRIES {
-            if let Some(min) = self.roots.iter().min().cloned() {
+            if let Some(min) = self.roots().min() {
                 self.roots.remove(&min);
-                self.cache.retain(|_, (fork, _, _)| *fork > min);
+                self.cache.retain(|_key, value| {
+                    let (max_slot, _, _) = &**value;
+                    max_slot.load(Ordering::Relaxed) > min
```

Comment:
noting for myself.

`retain` will grab the **write** lock on the shard for checking the closure. This guarantees that we have unique access to this `self.cache` entry.
Concurrent access from `insert` will have completed, or will be held up wiating for this to complete on the shard; if removed, then insert would be possibly re-adding entry (that's fine).

I'm like 99% sure that Relaxed ordering here (and in `insert`) is fine because we have guaranteed the unique access via the write-lock on shard.